Add shared support for MongoDB Dev Services container

patrox opened the pull request.

patrox committed:
Add shared MongoDB container for Dev Services

gsmet commented on extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/DevServicesMongoProcessor.java:
```diff
@@ -48,6 +50,18 @@ public class DevServicesMongoProcessor {
+     */
+    private static final String DEV_SERVICE_LABEL = "quarkus-dev-service-mongo";
+
+    private static final ContainerLocator mongoContainerLocator = new ContainerLocator(DEV_SERVICE_LABEL, MONGO_EXPOSED_PORT);
```
The constant should probably be named like a constant.

gsmet commented on extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/DevServicesBuildTimeConfig.java:
```diff
@@ -45,4 +45,30 @@ public class DevServicesBuildTimeConfig {
     public Map<String, String> containerEnv;
 
+    /**
+     * Indicates if the Mongo server managed by Quarkus Dev Services is shared.
```
Could you make sure you use `MongoDB` instead of `Mongo` in the javadoc of the configuration properties? They are published on the website.
(Note that other properties might not comply but let's not add more :))

patrox replied:
Sure, will fix it. Naming is so hard ;)

gsmet commented on extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/DevServicesMongoProcessor.java:
```diff
@@ -48,6 +50,18 @@ public class DevServicesMongoProcessor {
+     * Label to add to shared Dev Service for Mongo running in containers.
+     * This allows other applications to discover the running service and use it instead of starting a new instance.
+     */
+    private static final String DEV_SERVICE_LABEL = "quarkus-dev-service-mongo";
```
```suggestion
    private static final String DEV_SERVICE_LABEL = "quarkus-dev-service-mongodb";
```

gsmet commented on docs/src/main/asciidoc/mongodb-dev-services.adoc:
```diff
@@ -0,0 +1,30 @@
+
+If you need multiple (shared) servers, you can configure the `quarkus.mongo.devservices.service-name` attribute and indicate the server name.
+It looks for a container with the same value, or starts a new one if none can be found.
+The default service name is `mongo`.
```
Any chance we could make it `mongodb` or it is a convention and we should keep it that way?

patrox replied:
It can be changed

gsmet replied:
@patrox are you sure you pushed your changes? I don't see my comments addressed and also you said you squashed everything and the commits are not.

Thanks for your feedback!

patrox replied:
> @patrox are you sure you pushed your changes? I don't see my comments addressed and also you said you squashed everything and the commits are not.
> 
> Thanks for your feedback!

Yeah sorry, it looks like I messed things up during rebase. Now it should be fine :)

patrox replied:
@gsmet Now all requested changes should be in place, the commits were squashed and the branch is up-to-date with the current `main`, so it looks ready!

gsmet commented on docs/src/main/asciidoc/mongodb-dev-services.adoc:
```diff
@@ -0,0 +1,30 @@
+Most of the time you need to share the server between applications.
+Dev Services for MongoDB implements a _service discovery_ mechanism for your multiple Quarkus applications running in _dev_ mode to share a single server.
+
+NOTE: Dev Services for MongoDB starts the container with the `quarkus-dev-service-mongo` label which is used to identify the container.
```
```suggestion
NOTE: Dev Services for MongoDB starts the container with the `quarkus-dev-service-mongodb` label which is used to identify the container.
```

gsmet commented on docs/src/main/asciidoc/mongodb-dev-services.adoc:
```diff
@@ -0,0 +1,30 @@
+
+NOTE: Dev Services for MongoDB starts the container with the `quarkus-dev-service-mongo` label which is used to identify the container.
+
+If you need multiple (shared) servers, you can configure the `quarkus.mongo.devservices.service-name` attribute and indicate the server name.
```
I checked and I think the root config is `quarkus.mongodb` already?
```suggestion
If you need multiple (shared) servers, you can configure the `quarkus.mongodb.devservices.service-name` attribute and indicate the server name.
```

gsmet commented on docs/src/main/asciidoc/mongodb-dev-services.adoc:
```diff
@@ -0,0 +1,30 @@
+and pull requests should be submitted there:
+https://github.com/quarkusio/quarkus/tree/main/docs/src/main/asciidoc
+////
+= Dev Services for MongoDB (Configuration Free Databases)
```
Let's not clutter our guide index:
```suggestion
= Dev Services for MongoDB
```

gsmet reviewed (changes requested):
Thanks for this PR. It looks very nice.

I added a few minor gripes inline as to preferring `MongoDB`/`mongodb` over mongo. I'm not sure it makes sense in all circumstances so feel free to push back if it doesn't.

Could you also squash all your commits when done?

Thanks!

patrox commented:
> Thanks for this PR. It looks very nice.
> 
> I added a few minor gripes inline as to preferring `MongoDB`/`mongodb` over mongo. I'm not sure it makes sense in all circumstances so feel free to push back if it doesn't.
> 
> Could you also squash all your commits when done?
> 
> Thanks!

@gsmet I've addressed all your suggestions/comments and squashed all commits after rebasing to the current `main`.

gsmet committed:
Fix a small typo

gsmet reviewed (approved):
I amended your commit to fix `DEV_SERVICE_LABEL` that was still the `mongo` version.

LGTM, let's merge when CI is green.

Sorry it took so much time for me to get back to it :).

patrox commented:
> I amended your commit to fix `DEV_SERVICE_LABEL` that was still the `mongo` version.
> 
> LGTM, let's merge when CI is green.
> 
> Sorry it took so much time for me to get back to it :).

Thank you @gsmet, will wait for the CI to become 🟢

gsmet commented:
@patrox thanks for your contribution, it's going to be part of 3.12.
